core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -92,7 +92,10 @@ class DelayedFetch(
             // has just rolled, then the high watermark offset will remain the same but be on the old segment,
             // which would incorrectly be seen as an instance of Case F.
             if (endOffset.messageOffset != fetchOffset.messageOffset) {
-              if (endOffset.onOlderSegment(fetchOffset)) {
+              if (endOffset.messageOffsetOnly() || fetchOffset.messageOffsetOnly()) {
```

Comment:
I can understand the `endOffset.messageOffsetOnly()` case since the leader's high watermark is still not updated. But when will `fetchOffset` be `messageOffsetOnly`?

Reply:
`fetchOffset` can be message-only metadata when there is a diverging-epoch. If there is a diverged-epoch in the LogReadResults, then it won't enter the DelayedFetch. We can remove the check if it is not required.

Reply:
`fetchOffset` typically shouldn't be message only. But it doesn't hurt to have the check.